### ISSUE TITLE
Case splitting on earlier variable uses duplicate variable name

### ISSUE THREAD

**Issue by Snicksie** [agda/agda#2826]
Fairly minimal example:

```agda
open import Data.Nat

data Test {p} (P : Set p) : Set p where
  a : ( p : P) → Test P
  b : ( q : P) → Test P

doTest : (n : ℕ) → Test ℕ
doTest 0 = a 0
doTest (suc n) = b n

min-ex : (n : ℕ) → Test ℕ → Test ℕ
min-ex n s with doTest n
min-ex n s | a p = {!!}
min-ex n s | b q = {!!}
```

case splitting on s in the first hole gives:
```agda
min-ex : (n : ℕ) → Test ℕ → Test ℕ
min-ex n s with doTest n
min-ex n (a p) | a p = ?
min-ex n (b q) | a p = ?
min-ex n s | b q = {!!}
```

The variable p is now used twice, which is obviously not accepted by the type checker:

> Repeated variables in pattern: p
> when scope checking the left-hand side min-ex n (a p) in the
> definition of min-ex

If I'm first splitting on s and then on the result of doTest n, the result is correct as expected.

```agda
min-ex n s with doTest n
min-ex n (a p) | a p₁ = {!!}
min-ex n (a p) | b q = {!!}
min-ex n (b q) | t = {!!}
```

**Comment by sattlerc** [comment]
Shrunk test case:
```agda
data A : Set where
  a : A

data B : Set where
  b : (x : A) → B

foo : B → B
foo s with a
-- case splitting produces repeated variables in pattern
foo s | x = {!s!}
```

**Comment by andreasabel** [comment]
Works in Agda-2.5.1.1: 
```
foo (b x₁) | x = {!!}
```

**Comment by andreasabel** [comment]
My guess is that the regression was introduced by the change below.  The problem seems to be that with-patterns are not taken into consideration when assigning fresh variable by the logic below.
This problem might just disappear by the refactoring I do for #2822.

commit 67d50a9159147f0aa990aef4ae3ccbbfdd2b56df
Author:	Ulf Norell <ulfn@chalmers.se>  Tue Nov 29 16:57:44 2016
Committer:	Ulf Norell <ulfn@chalmers.se>  Tue Nov 29 16:57:44 2016

[ fixed #2018 ] don't rename variables when adding them to the context

... but make sure to pick fresh names when the system is choosing the name.

This means we have to pick fresh names for clauses generated by make case
in the AbstractToConcrete translation instead of relying on addContext to do it
for us. That makes more sense anyway since the old implementation used the
context of the left-hand side before pattern matching when picking fresh names.

**Comment by andreasabel** [comment]
> This problem might just disappear by the refactoring I do for #2822.

Turns out this was indeed the case.
